kanidmd/src/lib/modify.rs

```diff
@@ -133,6 +134,30 @@ impl ModifyList<ModifyInvalid> {
         }
     }
 
+    pub fn from_patch(
+        audit: &mut AuditScope,
+        pe: &ProtoEntry,
+        qs: &QueryServerWriteTransaction,
+    ) -> Result<Self, OperationError> {
+        let mut mods = Vec::new();
+
+        pe.attrs.iter().try_for_each(|(attr, vals)| {
+            // Issue a purge to the attr.
+            mods.push(m_purge(&attr));
+            // Now if there are vals, push those too.
+            // For each value we want to now be present.
+            vals.iter().try_for_each(|val| {
+                qs.clone_value(audit, &attr, &val).map(|resolved_v| {
+                    mods.push(Modify::Present(attr.as_str().into(), resolved_v));
+                })
+            })
+        })?;
+        Ok(ModifyList {
+            valid: ModifyInvalid,
+            mods,
+        })
+    }
+
```

Comment:
How does this function differ from others? 🤔

Reply:
patch is a diff that says 'only update the referenced fields'. Put says "replace the whole entity with the content of this entity". And "post" is create something new from this entity.

So it's different because the point of "patch" says "make these attributes, and only these ones look like this", so this enables us to transform that partial entry from the client into a change state request.